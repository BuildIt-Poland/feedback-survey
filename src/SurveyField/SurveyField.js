import React from 'react';
import styled from 'styled-components';
import { FormField, RadioButtonGroup, Select } from 'grommet';

import { spacingLarge } from '../styles/variables';

const Wrapper = styled.div`
  margin: 50rem ${spacingLarge};
`;

const renderImageOption = option => {
  return (
    <>
      <img width="50" src={'./image/' + option.toLowerCase() + '.svg'} alt={option} />
      <span>{option}</span>
    </>
  );
};

const buildAnswerOptions = (field, answerOptions) => {
  return answerOptions[field.answerOptions].map(option => {
    const label = field.type === 'rating' ? renderImageOption(option) : option;
    return {
      id: option + '_' + field.id,
      value: option,
      label
    };
  });
};

const renderFieldByType = (field, answerOptions) => {
  switch (field.type) {
    case 'open-ended': {
      return <FormField name={field.name} label={field.content} required={field.required} />;
    }
    case 'select': {
      return (
        <FormField
          name={field.name}
          label={field.content}
          required={field.required}
          component={Select}
          options={answerOptions[field.answerOptions]}
        />
      );
    }
    case 'radio': {
      return (
        <FormField
          name={field.name}
          label={field.content}
          required={field.required}
          component={RadioButtonGroup}
          options={buildAnswerOptions(field, answerOptions)}
        />
      );
    }
    case 'rating': {
      return (
        <FormField
          name={field.name}
          label={field.content}
          required={true}
          component={RadioButtonGroup}
          options={buildAnswerOptions(field, answerOptions)}
        />
      );
    }
    default: {
      return null;
    }
  }
};

const SurveyField = function({ field, answerOptions }) {
  if (!field) {
    return null;
  }
  return <Wrapper>{renderFieldByType(field, answerOptions)}</Wrapper>;
};

export default SurveyField;
