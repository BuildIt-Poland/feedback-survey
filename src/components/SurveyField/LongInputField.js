import React from 'react';
import { FormField, TextArea } from 'grommet';

const LonfInputField = function({ field }) {
  if (!field) {
    return null;
  }
  return (
    <FormField
      name={field.id.toString()}
      label={field.content}
      required={field.required}
      component={TextArea}
      placeholder="Type your answer here"
    />
  );
};

export default LonfInputField;
