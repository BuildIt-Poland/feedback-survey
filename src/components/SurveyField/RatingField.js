import React from 'react';
import PropTypes from 'prop-types';
import { FormField } from 'grommet';

import { formField, radioOption } from '../../types';
import CustomRadioButtonGroup from '../CustomRadioButtonGroup';

const RatingField = ({ field, options }) => {
  return (
    <FormField
      name={field.id}
      label={field.content}
      required={field.required}
      component={CustomRadioButtonGroup}
      options={options}
      direction="row-responsive"
      width="48rem"
      withIcon={true}
    />
  );
};

RatingField.propTypes = {
  field: formField.isRequired,
  options: PropTypes.arrayOf(radioOption)
};

export default RatingField;
