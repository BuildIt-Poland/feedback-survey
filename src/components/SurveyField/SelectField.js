import React from 'react';
import { FormField } from 'grommet';

import CustomSelect from '../CustomSelect';

const SelectField = function({ field, options }) {
  if (!field) {
    return null;
  }
  return (
    <FormField
      name={field.id.toString()}
      label={field.content}
      required={field.required}
      component={CustomSelect}
      options={options}
      placeholder="Select"
    />
  );
};

export default SelectField;
