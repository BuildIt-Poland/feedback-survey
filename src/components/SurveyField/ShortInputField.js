import React from 'react';
import { FormField } from 'grommet';

import { fontWeightNormal, colorWhite } from '../../styles/variables';
import { formField } from '../../types';

export const inputTheme = {
  global: {
    input: {
      weight: fontWeightNormal
    },
    focus: {
      border: {
        color: colorWhite
      }
    }
  }
};

const ShortInputField = ({ field }) => {
  return (
    <FormField name={field.id} label={field.content} required={field.required} placeholder="Type your answer here" />
  );
};

ShortInputField.propTypes = {
  field: formField.isRequired
};

export default ShortInputField;
