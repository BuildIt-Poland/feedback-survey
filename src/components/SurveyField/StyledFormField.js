import styled from 'styled-components';
import { FormField } from 'grommet';

import media from '../../styles/media';
import { spacingMedium, fontMedium, fontSmall } from '../../styles/designTokens';

const StyledFormField = styled(FormField)`
  max-width: 80rem;
  div {
    ${props => (props.noBorderBottom ? 'border-bottom: none' : null)}
  }
  input {
    font-weight: 400;

    :focus {
      border-color: red;
    }
  }
  label:first-child {
    margin: 0 0 2rem 0;
    font-size: ${fontSmall};

    &:after {
      color: red;
      content: ${props => (props.required ? "'*'" : "''")};
      margin-left: 0.5rem;
      white-space: nowrap;
    }
  }

  ${media.phone`
    label:first-child {
      font-size: ${fontMedium};
    }
  `}
`;

export default StyledFormField;
