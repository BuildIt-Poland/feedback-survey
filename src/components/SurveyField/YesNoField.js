import React from 'react';
import PropTypes from 'prop-types';
import { FormField } from 'grommet';

import { formField, radioOption } from '../../types';
import CustomRadioButtonGroup from '../CustomRadioButtonGroup';

const YesNoField = ({ field, options }) => {
  return (
    <FormField
      name={field.id}
      label={field.content}
      required={field.required}
      component={CustomRadioButtonGroup}
      options={options}
      width="medium"
      withIcon={false}
    />
  );
};

YesNoField.propTypes = {
  field: formField.isRequired,
  options: PropTypes.arrayOf(radioOption)
};

export default YesNoField;
