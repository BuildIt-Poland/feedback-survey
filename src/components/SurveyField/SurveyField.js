import React from 'react';
import styled from 'styled-components';
import { arrayOf } from 'prop-types';
import { Grommet } from 'grommet';

import { spacingLarge, fontMedium, spacingMedium, colorRed, colorBlue } from '../../styles/designTokens';
import { formField, answerType } from '../../types';
import RatingField from './RatingField';
import ShortInputField from './ShortInputField';
import LongInputField from './LongInputField';
import SelectField from './SelectField';
import YesNoField from './YesNoField';

const Wrapper = styled.div`
  margin: 5rem ${spacingLarge};
  position: relative;

  span[data-type='required-field'] {
    position: absolute;
    top: 0.3rem;
    color: ${colorRed};
  }
`;

const theme = {
  formField: {
    extend: {
      label: {
        'font-size': fontMedium,
        'margin-bottom': spacingMedium
      }
    }
  },
  select: {
    icons: {
      color: colorBlue
    }
  },
  global: {
    colors: {
      focus: colorBlue,
      brand: colorBlue
    },
    input: {
      weight: '400'
    },
    focus: {
      border: {
        color: colorBlue
      }
    }
  }
};

const getAnswerTypes = (array, answerType) => array.find(item => item.type === answerType).values;

const buildAnswerTypes = ({ answerType, type, id }, answerTypes) => {
  const types = getAnswerTypes(answerTypes, answerType);
  return types.map(option => {
    return {
      id: `${option}_${id}`,
      value: option,
      label: option
    };
  });
};

const renderFieldByType = (field, answerTypes, value, error, onChange, setFieldValue) => {
  switch (field.type) {
    case 'open-ended': {
      return <LongInputField field={field} error={error} value={value} onChange={onChange} />;
    }
    case 'select': {
      return (
        <SelectField
          field={field}
          options={getAnswerTypes(answerTypes, field.answerType)}
          error={error}
          value={value}
          setFieldValue={setFieldValue}
        />
      );
    }
    case 'radio': {
      return (
        <YesNoField
          field={field}
          options={buildAnswerTypes(field, answerTypes)}
          error={error}
          value={value}
          onChange={onChange}
        />
      );
    }
    case 'rating': {
      return (
        <RatingField
          field={field}
          options={buildAnswerTypes(field, answerTypes)}
          error={error}
          value={value}
          onChange={onChange}
        />
      );
    }
    default: {
      return <ShortInputField field={field} error={error} value={value} onChange={onChange} />;
    }
  }
};

const SurveyField = ({ field, answerTypes, value, error, onChange, setFieldValue }) => (
  <Grommet theme={theme}>
    <Wrapper>{renderFieldByType(field, answerTypes, value, error, onChange, setFieldValue)}</Wrapper>
  </Grommet>
);

SurveyField.propTypes = {
  field: formField.isRequired,
  answerTypes: arrayOf(answerType)
};

export default SurveyField;
