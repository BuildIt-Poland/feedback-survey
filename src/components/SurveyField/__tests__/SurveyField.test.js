import React from 'react';
import { create } from 'react-test-renderer';

import SurveyField from '../SurveyField';

describe('COMPONENT - SurveyField', () => {
  it('renders correctly', () => {
    const field = {
      id: '123',
      content: 'Question content',
      required: false,
      options: ['Bad', 'Average'],
      type: 'kindOfFeedback'
    };
    const answerTypes = [
      {
        type: 'kindOfFeedback',
        values: ['monthly', 'quaterly', 'annual']
      }
    ];
    const component = create(<SurveyField field={field} answerTypes={answerTypes} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
