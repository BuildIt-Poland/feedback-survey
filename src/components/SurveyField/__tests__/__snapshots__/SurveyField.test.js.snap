// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - SurveyField renders correctly 1`] = `
<div
  className="StyledGrommet-sc-19lkkz7-0 eUfLDp"
>
  <div
    className="sc-bxivhb fpvsvs"
  >
    <div
      className="FormField__FormFieldBox-m9hood-0 OcTuP StyledBox-sc-13pk1d4-0 CAsMy"
    >
      <label
        className="StyledText-sc-1sadyjn-0 bMXxti"
      >
        Question content
      </label>
      <div
        className="StyledBox-sc-13pk1d4-0 gZSidW"
      />
    </div>
  </div>
</div>
`;
